crates/jstzd/src/container.rs

```diff
@@ -0,0 +1,114 @@
+use async_dropper::{AsyncDrop, AsyncDropper};
+use async_trait::async_trait;
+use bollard::{container::StartContainerOptions, Docker};
+use std::sync::Arc;
+
+use crate::network::Network;
+
+pub type ContainerId = String;
+#[derive(Default)]
+struct ContainerInner {
+    pub id: ContainerId,
+    pub name: String,
+    client: Option<Arc<Docker>>,
+    // TODO: network: Option<Network>,
+    _private: (),
+}
+
+pub struct Container(AsyncDropper<ContainerInner>);
+
+impl ContainerInner {
+    /// Creates a new container with running `id`
+    pub fn new(
+        client: Arc<Docker>,
+        id: ContainerId,
+        name: &str,
+        _network: Option<Network>,
+    ) -> Self {
+        Self {
+            id,
+            name: name.to_string(),
+            client: Some(client),
+            // network,
+            _private: (),
+        }
+    }
+
+    /// Starts the container's entrypoint
+    pub async fn start(&self) -> anyhow::Result<()> {
+        let client = self.client()?;
+        if let Some(state) = client.inspect_container(&self.name, None).await?.state {
+            if let Some(true) = state.running {
+                println!("Container is already running");
+                return Ok(());
+            }
+        }
+        client
+            .start_container(&self.name, None::<StartContainerOptions<String>>)
+            .await?;
+        Ok(())
+    }
+
+    /// Stop the container
+    pub async fn stop(&self) -> anyhow::Result<()> {
+        self.client()?.stop_container(&self.name, None).await?;
+        Ok(())
+    }
+
+    fn client(&self) -> anyhow::Result<&Arc<Docker>> {
+        self.client
+            .as_ref()
+            .ok_or(anyhow::anyhow!("client not found"))
+    }
+}
+
+#[async_trait]
+impl AsyncDrop for ContainerInner {
+    async fn async_drop(&mut self) {
+        let client = self.client().expect("client not found");
+        self.stop()
+            .await
+            .unwrap_or_else(|e| eprintln!("Error stopping container: {}", e));
+        match client.remove_container(&self.name, None).await {
+            Ok(_) => {}
+            Err(e) => eprintln!("Error removing container {}: {}", self.name, e),
+        }
+        // TODO: Handle network drop?
+    }
+}
+
+impl Container {
+    pub fn new(
+        client: Arc<Docker>,
+        id: ContainerId,
+        name: &str,
+        network: Option<Network>,
+    ) -> Self {
+        Self(AsyncDropper::new(ContainerInner::new(
+            client, id, name, network,
+        )))
+    }
+
+    pub fn id(&self) -> &ContainerId {
+        &self.0.inner().id
+    }
+
+    pub fn name(&self) -> &str {
+        &self.0.inner().name
+    }
+
+    pub async fn start(&self) -> anyhow::Result<()> {
+        self.0.start().await
+    }
+
+    pub async fn stop(&self) -> anyhow::Result<()> {
+        self.0.stop().await
+    }
+}
```

Comment:
Suggest using `#[derive(AsyncDrop)]` on `ContainerInner` (to be renamed to `Container`) to avoid this wrapper boilerplate

Reply:
can't use the `derive` feature together w the `simple` feature that enables us to remove the Default trait. I will stick to the simple feature for now